File: origLP.py
from dataclasses import dataclass
import numpy as np

@dataclass
class event_t:
    x: np.uint16    # x position of event
    y: np.uint16    # y position of event
    t: int          # event time in us
    p: int          # polarity of event (1 == on, 0 == off)

@dataclass
class flow_event_t:
    tc: event_t
    vx: float
    vy: float
    mag: float

# Local Planes algorithm
class origLP:
    def __init__(self, frame_height, frame_width, dt):
        self.frame_height = frame_height
        self.frame_width = frame_width
        self.dt = dt
        frame_size = frame_height * frame_width
        self.on_event_cloud = np.zeros(frame_size)
        self.off_event_cloud = np.zeros(frame_size)
        self.total_evts = 0
        self.maxDtThreshold = dt*5
        self.time = 0

    def update_time(self):
        self.time += self.dt

    def process_frame(self, tframe):
        flow_events = []

        self.update_event_clouds(tframe)
        
        for i in range(self.frame_height):
            for j in range(self.frame_width):
                if tframe[i,j] != 0:
                    if tframe[i,j] == 1:
                        event = event_t(i,j,self.time,1)
                    elif tframe[i,j] == -1:
                        event = event_t(i,j,self.time,-1)
                    flow_events.append(self.get_flow(event))

        self.update_time()

        return flow_events
    
    def update_event_clouds(self, tframe):
        for i in range(self.frame_height):
            for j in range(self.frame_width):
                if tframe[i,j] != 0:
                    if tframe[i,j] == 1:
                        self.on_event_cloud[i * self.frame_width + j] = self.time
                    elif tframe[i,j] == -1:
                        self.off_event_cloud[i * self.frame_width + j] = self.time
                

    def get_flow(self, event: event_t):
        # add event to event cloud
        # last = 0
        # if event.p == 1:
        #     last = self.on_event_cloud[event.x * self.frame_width + event.y]
        #     self.on_event_cloud[event.x * self.frame_width + event.y] = event.t
        # else:
        #     last = self.off_event_cloud[event.x * self.frame_width + event.y]
        #     self.off_event_cloud[event.x * self.frame_width + event.y] = event.t

        # maybe add check if pixel has been inactive for at least t_refract
        lf = self.compute_local_plane(event, 5)
        self.total_evts += 1
        if lf.mag != 0:
            return lf
        else:
            return flow_event_t(event, 0, 0, 0)

    def compute_local_plane(self, event: event_t, searchDistance: int):
        if event.p == 1:
            event_cloud = self.on_event_cloud
        else:
            event_cloud = self.off_event_cloud
        
        a10 = 0
        a01 = 0
        ii = 0
        jj = 0

        # idk what these variable values should be
        cut = 1
        thr = 1e-9

        # search area along x axis
        for i in range(-searchDistance, searchDistance+1):
            # search area along y axis
            for j in range(-searchDistance, searchDistance+1):
                # check if pixel is in frame
                if (event.x+1 >= 0 and event.x+i < self.frame_height and event.y+j >= 0 and event.y+j < self.frame_width):
                    # get event for current pixel
                    t1 = event_cloud[(event.x + i) * self.frame_width + (event.y + j)]

                    # check if event exists and if event is recent enough
                    if (t1 != 0 and event.t-t1 < self.maxDtThreshold):
                        # search for nearby events along x axis
                        for xx in range(i+1, searchDistance+1):
                            if (event.x+xx >= 0 and event.x+xx < self.frame_height): # check it's in frame
                                t2 = event_cloud[(event.x + xx) * self.frame_width + (event.y + j)]
                                if (t2 != 0 and event.t-t2 < self.maxDtThreshold):
                                    a10 += float(t2-t1)/(xx-i)
                                    ii += 1

                        # search for nearby events along y axis
                        for yy in range(j+1, searchDistance+1):
                            if (event.y+yy >= 0 and event.y+yy < self.frame_width): # check it's in frame
                                t2 = event_cloud[(event.x + i) * self.frame_width + (event.y + yy)]
                                if (t2 != 0 and event.t-t2 < self.maxDtThreshold):
                                    a01 += float(t2-t1)/(yy-j)
                                    jj += 1
        
        if (ii < cut or jj < cut):
            return flow_event_t(event, 0, 0, 0)
        else:
            a10 /= ii
            a01 /= jj

        # a10 *= 1e-6
        # a01 *= 1e-6

        if (abs(a10) < thr and abs(a01) < thr):
            return flow_event_t(event, 0, 0, 0)
        else:
            temp = 1.0 / (a10 * a10 + a01 * a01)
            vx = a10 * temp
            vy = a01 * temp
            mag = round(np.sqrt(vx * vx + vy * vy))
            return flow_event_t(event, vx, vy, mag)

File: origLP.cpp
/**
 * Local Plane with Savitzky Golay filter from https://www.frontiersin.org/articles/10.3389/fnins.2016.00176/full
 */

#include "origLP.hpp"

#include <stdlib.h>
#include <math.h>

origLP::origLP()
{
    total_evts = 0;
}

origLP::~origLP()
{
	delete[] on_event_cloud;
    delete[] off_event_cloud;
    // swfile.close();
}

void origLP::init()
{
	// swfile.open("sw_results.txt");

	// initialize local flow data arrays
    on_event_cloud = new int[frame_length];
    off_event_cloud = new int[frame_length];

    for (int i = 0; i < frame_length; ++i)
    {
        on_event_cloud[i] = 0;
        off_event_cloud[i] = 0;
    }
}

flow_event_t origLP::get_flow(event_t event)
{
	// add event to event cloud
    int last = 0;
    if (event.p == 1)
    {
        last = on_event_cloud[event.x * frame_width + event.y];
        on_event_cloud[event.x * frame_width + event.y] = event.t;
    }
    else
    {
        last = off_event_cloud[event.x * frame_width + event.y];
        off_event_cloud[event.x * frame_width + event.y] = event.t;
    }

    // check if pixel has been inactive for at least t_refract
    // if (last + refractTime <= event.t)
    // {
        // compute local, normal flow
        flow_event_t lf = _compute_local_plane(event);
        total_evts++;
        if (lf.mag != 0)
        {
            return lf;
        }
        else
            return flow_event_t(event, 0, 0, 0);
    // }
    // else
    //     return flow_event_t(event, 0, 0, 0);
}


flow_event_t origLP::_compute_local_plane(event_t event)
{
	int *event_cloud;
    if (event.p == 1)
        event_cloud = on_event_cloud;
    else
        event_cloud = off_event_cloud;

    float a10 = 0;
    float a01 = 0;
    int ii = 0;
    int jj = 0;

    for (int i = -searchDistance; i <= searchDistance; ++i)
    {
        for (int j = -searchDistance; j <= searchDistance; ++j)
        {
            if (event.x + i >= 0 && event.x + i < frame_height && event.y + j >= 0 && event.y + j < frame_width) // within bounds
            {
                int t1 = event_cloud[(event.x + i) * frame_width + (event.y + j)]; // pixel in frame
                if (t1 != 0 && event.t - t1 < maxDtThreshold)
                {
                    for (int xx = i + 1; xx <= searchDistance; ++xx)
                    {
                        if (event.x + xx >= 0 && event.x + xx < frame_height)
                        {
                            int t2 = event_cloud[(event.x + xx) * frame_width + (event.y + j)];
                            if (t2 != 0 && event.t - t2 < maxDtThreshold)
                            {
                                a10 += (float)(t2 - t1) / (xx - i);
                                ii++;
                            }
                        }
                    }
                    for (int yy = j + 1; yy <= searchDistance; ++yy)
                    {
                        if (event.y + yy >= 0 && event.y + yy < frame_width)
                        {
                            int t2 = event_cloud[(event.x + i) * frame_width + (event.y + yy)];
                            if (t2 != 0 && event.t - t2 < maxDtThreshold)
                            {
                                a01 += (float)(t2 - t1) / (yy - j);
                                jj++;
                            }
                        }
                    }
                }
            }
        }
    }

    swfile << a10 << " " << a01 << std::endl;

    if (ii < cut || jj < cut)
        return flow_event_t(event, 0, 0, 0);
    else
    {
        a10 /= ii;
        a01 /= jj;
    }

    a10 *= 1e-6;
    a01 *= 1e-6;

    if (std::abs(a10) < thr && std::abs(a01) < thr)
        return flow_event_t(event, 0, 0, 0);
    else
    {
        float temp = 1.0 / (a10 * a10 + a01 * a01);
        float vx = a10 * temp;
        float vy = a01 * temp;
        float mag = std::round(std::sqrt(vx * vx + vy * vy));
        return flow_event_t(event, vx, vy, mag);
    }
}